### DISCUSSION TITLE
[RFC] New Component: Splitter

### DISCUSSION THREAD

**Discussion by wanpan11** [ant-design/ant-design#49924] [RFCs - archive]
## Motivation

分割面板用于隔离区域，展示多个内容.

## Components

- Splitter
  - 对外暴露组件
- Panel
  - 对外暴露面板组件
- SplitBar
  - 操作组件

### Splitter

| Props         | Descriptions     | Type                        | Default      |
| ------------- | ---------------- | --------------------------- | ------------ |
| layout        | 布局方向         | `horizontal` \| `vertical`  | `horizontal` |
| style         | 容器样式         | `css-properties`            | -            |
| onResizeStart | 开始拖拽之前回调 | `(sizes: number[]) => void` | -            |
| onResize      | 面板大小变化回调 | `(sizes: number[]) => void` | -            |
| onResizeEnd   | 拖拽结束回调     | `(sizes: number[]) => void` | -            |

### Panel

| Props       | Descriptions                                  | Type               | Default |
| ----------- | --------------------------------------------- | ------------------ | ------- |
| min         | 最小阈值 `1-100 \| '10%' \| '200px'`          | `number \| string` | -       |
| max         | 最大阈值 `1-100 \| '10%' \| '200px'`          | `number \| string` | -       |
| size        | 受控面板大小 `1-100 \| '10%' \| '200px'`      | `number`           | -       |
| defaultSize | 初始面板大小 `1-100 \| '10%' \| '200px'`      | `number`           | -       |
| collapsible | 快速折叠 `collapsible=true`时将忽略`min``max` | `boolean`          | `false` |
| resizable   | 是否支持拖拽伸缩                              | `boolean`          | `true`  |

## Token

| Props               | Descriptions           | Type     | Default |
| ------------------- | ---------------------- | -------- | ------- |
| resizableSize       | 可改变大小标识元素大小 | `number` | `10`    |
| collapsibleIconSize | 快速折叠图标大小       | `number` | `12`    |

## Example

[preview](https://github.com/ant-design/ant-design/pull/50038#issuecomment-2246752430)

```ts
const App: React.FC = () => (
  <Splitter
    layout={layout}
    style={{
      height: 300,
      borderRadius: '4px',
      boxShadow: '0 0 10px rgba(0, 0, 0, 0.1)',
    }}
  >
    <Splitter.Panel defaultSize="220px" collapsible>
      <Card title="first" bordered={false}>
        <div>collapsible: true</div>
      </Card>
    </Splitter.Panel>

    <Splitter.Panel>
      <Card title="second" bordered={false}>
        <div>something</div>
      </Card>
    </Splitter.Panel>
  </Splitter>
);
```

<img width="1192" alt="image" src="https://github.com/user-attachments/assets/17da525c-77f9-4856-a190-a207f3670ec6">

**Comment by MadCcc** [comment]
参考一下 https://github.com/ant-design/ant-design/discussions/48374
API 设计，有哪些组件，期望的用法都应该列出来

**Comment by MadCcc** [comment]
用法参考
https://arco.design/react/components/resize-box
https://pro-editor.antdigital.dev/components/draggable-panel
https://devui.design/components/zh-cn/splitter/demo

**Comment by wanpan11** [comment]
参考
https://primereact.org/splitter
https://ui.shadcn.com/docs/components/resizable

**Comment by huangkairan** [comment]
现在在用 https://ui.lobehub.com/components/draggable-panel 基本满足需求

**Comment by MadCcc** [comment]
1. 组件名：Splitter
子组件：Splitter.Panel
使用 DOM 结构
```
<Splitter>
  <Splitter.Panel>
</Splitter>
```
2. API
- height 不需要，使用 style 透传给容器
- splitBarSize 暂时不提供
- min/max 需要考虑百分比（string）
- size，默认属性可以叫 `defaultSize`，需要增加受控方法 `onResize`

**Reply by MadCcc** [reply]
3. 考虑实现合并拖拽

**Reply by wanpan11** [reply]
佬 这个用 DOM 结构是有撒考虑吗
个人感觉 Splitter.Panel 这个作为单独组件暴露出去没撒作用 
用 items 去接受感觉也不错🤔

**Reply by MadCcc** [reply]
作为 layout 组件使用 DOM 结构更符合直觉，比如 Layout.Sider 等。

**Comment by MadCcc** [comment]
设计需要确认的点
1. 是否有形态变体（有无边框）
2. 拖动条的交互效果

**Comment by JSchmidt63** [comment]
see SplitBox at antvis X6:
https://github.com/antvis/X6/tree/v1
https://github.com/antvis/X6/tree/v1/packages/x6-react-components
https://github.com/antvis/X6/tree/v1/packages/x6-react-components/src/split-box

demo at:
https://x6.antv.vision/zh/docs/api/ui/splitbox

**Comment by wanpan11** [comment]
- [x] height 不需要，使用 style 透传给容器
- [x] splitBarSize 暂时不提供
- [x] min/max 需要考虑百分比（string）
- [x] size，默认属性可以叫 defaultSize，需要增加受控方法 onResize
- [x] 使用 DOM 结构
